File: Top 100 Codes/02-Working with Numbers/18-Count possible decoding of a given digit sequence/Countpossibledecodingofagivendigitsequence.py
def count_decodings(digits, length):
    """
    Function to count possible decodings of a given digit sequence.

    :param digits: String representing the digit sequence
    :param length: Length of the digit sequence
    :return: Total number of possible decodings
    """

    # Edge case: If the sequence starts with '0', decoding is not possible
    if length == 0 or digits[0] == '0':
        return 0

    # Array to store decoding counts for substrings
    count = [0] * (length + 1)
    
    # Base cases
    count[0], count[1] = 1, 1

    for k in range(2, length + 1):
        count[k] = 0  # Initialize current position

        # If the last digit is not '0', carry over the previous count
        if digits[k - 1] > '0':
            count[k] = count[k - 1]

        # If the last two digits form a valid number (10-26), add the count from two steps back
        if digits[k - 2] == '1' or (digits[k - 2] == '2' and digits[k - 1] < '7'):
            count[k] += count[k - 2]

    return count[length]


# Take user input
digits = input("Enter the sequence: ")

# Display the result
print("Possible count of decodings:", count_decodings(digits, len(digits)))

File: Top 100 Codes/02-Working with Numbers/18-Count possible decoding of a given digit sequence/Countpossibledecodingofagivendigitsequence.cpp
#include <iostream>
#include <cstring> // For strlen()
using namespace std;

// Function to count possible decodings of a given digit sequence
int countDecodings(char *digits, int length) {
    // If the input sequence starts with '0', it can't be decoded
    if (length == 0 || digits[0] == '0')
        return 0;

    // Array to store the count of decodings up to each position
    int count[length + 1];
    count[0] = 1; // Empty string has one way to decode (do nothing)
    count[1] = 1; // Single non-zero digit has one way to decode

    for (int i = 2; i <= length; i++) {
        count[i] = 0; // Initialize the current count to zero

        // If the last digit is not '0', it contributes to decoding
        if (digits[i - 1] > '0')
            count[i] = count[i - 1];

        // If the last two digits form a valid number (10-26), include those
        if (digits[i - 2] == '1' || (digits[i - 2] == '2' && digits[i - 1] < '7'))
            count[i] += count[i - 2];
    }
    return count[length];
}

int main() {
    char digits[15];

    // Take input from the user
    cout << "Enter the sequence: ";
    cin >> digits;

    int length = strlen(digits);

    // Display the result
    cout << "Possible count of decodings: " << countDecodings(digits, length);

    return 0;
}
